### PR TITLE
Fix page timing issues

### PR DESCRIPTION
**Pull request by aweng98**
- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/FireCloudView.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/FireCloudView.scala]: Nit: I'm not sure what's going on with your editor but there's no reason for these import changes, which I'd actually say are now in the wrong order.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Clickable.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Clickable.scala:17]: I think we want to continue doing these 3 things here in `Clickable` where we know what conditions must be met before clicking on something. We don't want to rely on every subtype's implementation of `awaitReady` to do the right thing.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Component.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Component.scala]: I don't think we should be scrolling the view in `awaitReady`. From what I've seen, the only reason we use `scrollToVisible` at all is for clicking on things that may be off screen... and that has only been necessary with more recent versions of selenium (and may even be a bug that has been fixed and this is therefore not needed anymore).

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/PageUtil.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/PageUtil.scala]: As discussed in person, please include some comments here about the difference between the two ways to navigate and why we're trying this experiment.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/user/SignInPage.scala]: Seeing the `isOpen` function on `Modal`, I kinda like that name for pages too; it matches the `open` function.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala]: This probably isn't a problem right now, but this won't always hold if another tab is loaded. Just need to watch out for calling `awaitReady` at times other than when the page is first opened.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala]: We should have a function on `Button` that encapsulates these two checks.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala]: Style nit: this should be on the same line as the `{` above. Also, there might be another way to reference the containing page... I just can't think of it off the top of my head.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala:61]: Some of the things below open modals. Wouldn't it be more correct to check for those modals than to just check for the parent page being ready?

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala:106]: The more I see these, the more I think it's not right for sub components to delegate their ready state to the containing page. _Actions_ could result in the need to await for the page but sub components' state should be local to that component.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/FreeTrialSpec.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/FreeTrialSpec.scala]: Good catch. This should definitely happen (proactively) after a test, not just (defensively) before a test.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/FreeTrialSpec.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/FreeTrialSpec.scala]: Either this doesn't need to be explicitly specified or `managerAuthToken` above doesn't need to be `implicit`. I'm not familiar enough with this test to know, but we should decide which token makes the most sense to be implicit and make that one, and no others, implicit. (I'm making a guess here that you're passing it explicitly because the compiler complained that there were too many implicit candidates.)

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Component.scala
Comment (MattPutnam) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Component.scala]: I don't think scrollToVisible should be part of awaitReady

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/FireCloudView.scala
Comment (MattPutnam) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/FireCloudView.scala:12]: Why?

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Modal.scala
Comment (MattPutnam) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Modal.scala]: This isn't correct. This is checking if *any* modal is open. We can tell if a particular modal is open via `modal.isVisible`

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/TabBar.scala
Comment (MattPutnam) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/TabBar.scala]: I don't think this is correct. The tabs themselves don't have any children.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/PageUtil.scala
Comment (MattPutnam) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/PageUtil.scala]: Why?

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala
Comment (MattPutnam) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala]: We don't need to be testing the state of the tab bar. This page class shouldn't know anything about the tab bar, that's WorkspacePage's responsibility.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala
Comment (MattPutnam) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala]: I don't like this. We sometimes come to this page when we expect this button to stay disabled.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala
Comment (MattPutnam) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala]: I don't like the idea of creating a separate page just to borrow its url.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (MattPutnam) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala]: Unnecessary

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (MattPutnam) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala]: Can just use `awaitState("ready")`

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/FreeTrialSpec.scala
Comment (MattPutnam) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/FreeTrialSpec.scala]: Why has this been deleted from all of these tests? Also, this one needs to get moved to the top.

**Review by MattPutnam — CHANGES_REQUESTED** [review]

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/FireCloudView.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/FireCloudView.scala:12]: isn't this already defined in Awaiter?
```
trait Awaiter {
  def awaitReady(): Unit
}
```

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/TabBar.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/TabBar.scala]: inspecting DOM - when a tab is selected, two child divs rendered. only opened tab has child divs. this works.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Modal.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Modal.scala]: good catch!

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/PageUtil.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/PageUtil.scala]: for this reason (a theory):
The driver.get() method is handled by the browser as a fragment identifier.
The navigate().to() method is handled by the browser as a address/location/URL bar input.

**Review by breilly2 — CHANGES_REQUESTED** [review]

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/FreeTrialSpec.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/user/FreeTrialSpec.scala]: oh this is not suppose be included. I will revert changes.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/FireCloudView.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/FireCloudView.scala]: I blame this on Intellij. I don't know how to tell it NOT do it.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/PageUtil.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/PageUtil.scala]: I need to re-find the source. I will revert this change so PR is not blocked by it.

**Comment by aweng98** [comment]
FreeTrialSpec change wasn't suppose to be included. It is reverted back.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala]: to remove it, how do I then access outer `awaitReady` from inside of sidebar?

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala:106]: I think when parent component is "ready" or "enabled", all sub components are "ready" or "enabled". If one sub-component is ready but while other sub-components are not, it could makes UI to move or jitter. negative side-effect chain reaction.

Comment (breilly2) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala:106]: Interesting point. However it makes sense for a page's ready to be implemented by delegating down to the sub component's ready. If the sub component invokes the parent ready then it'll be an endless loop. The checks have to be one way. I think it's then up to the previous action to know how wide the await needs to be... either just a sub component or the whole page.

**Comment by aweng98** [comment]
jenkins retest

**Comment by aweng98** [comment]
jenkin slave out of space.
jenkins retest

**Comment by aweng98** [comment]
jenkins retest

**Comment by aweng98** [comment]
jenkins retest

**Comment by aweng98** [comment]
jenkins retest

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/TabBar.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/TabBar.scala]: new method is now removed. I need some time to come up with a better way handling check.

**Comment by aweng98** [comment]
jenkins retest

**Comment by aweng98** [comment]
jenkins retest

**Comment by aweng98** [comment]
jenkins retest

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala
Comment (MattPutnam) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala]: Remove

**Review by MattPutnam — APPROVED** [review]

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Component.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Component.scala]: `Clickable`s need to be enabled. I don't think that `Component`s in general should make this check.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala
Comment (breilly2) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala]: Yeah, it's not ideal. You could extract a function into a companion object:
```
object WorkspaceMethodConfigListPage {
  def url(namespace: String, name: String): String = s"${Config.FireCloud.baseUrl}#workspaces/$namespace/$name/method-configs"
}
```
to enable reuse of that logic while avoiding extra object instantiation.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (breilly2) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala:106]: I still really dislike the parent references. Awaits should delegate down, not up.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala:114]: This is `awaitReady` on the `Collapse`? I wouldn't expect that clicking this button affects the `Collapse` component itself, rather the content inside the `Collapse`. Looks like we could wait on the key input with a `data-test-id` of `-key`.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala]:
```
val aclEditor = sidebar.clickShareWorkspaceButton()
```

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala]:
```
val aclEditor = sidebar.clickShareWorkspaceButton()
```

**Review by breilly2 — CHANGES_REQUESTED** [review]
Almost there. Mostly nits, however I still have a problem with the components awaiting their parents. I think this is the wrong direction to be delegating.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala:114]: no, it's not `awaitReady` from `Collapse`. It's calling `workspaceAttributesArea.awaitReady` defined at line# 104.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala]: changed

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/summary/WorkspaceSummaryPage.scala]: changed

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/page/workspaces/methodconfigs/WorkspaceMethodConfigDetailsPage.scala]: done

**Comment by aweng98** [comment]
reverted WorkspaceMethodConfigDetailsPage.scala

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Component.scala
Comment (aweng98) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/component/Component.scala]: changed back

**Review by breilly2 — APPROVED** [review]
👍 Approving so we can thaw develop. However, I want to quickly (immediately) continue the discussion about how to structure awaits of pages and subcomponents.

**Comment by aweng98** [comment]
I'll schedule a meeting to discuss your concern among ui automation
stakeholders. in meantime, we can also observe how well jenkin jobs go for
next couple days.

thank you
Alex

**Pull request merged**
